### ISSUE TITLE
Performance issue with WebGPURenderer in Windows Chrome

### ISSUE THREAD

**Issue by danrossi** [mrdoob/three.js#26266]
### Description

I thought I would try here first as my AB vanilla test produces less cpu resources compared to WebGPURenderer. I am not sure what it is doing but in Chrome I have consistent cpu usage of 20% and 3% gpu. In Canary its 5-10% cpu and 20% gpu. In the Vanilla example it is  1-2% cpu for chrome production and Canary.  Hopefully the shader used in the vanilla example is relevant. 

### Reproduction steps

Updated WebGPURenderer

### Code

Updated texture with

```
texture.colorSpace = THREE.SRGBColorSpace;
```

### Live example

* [WebGPURenderer with colorspace fix.](https://danrossi.github.io/three-webgpu-renderer/tests/webgpu_video_panorama_equirectangular.html)
* [Vanilla example.]{https://danrossi.github.io/three-webgpu-renderer/tests/webgpu_basic.html
}

### Version

153dev

### Browser

Chrome

### OS

Windows

**Comment by danrossi** [comment]
I've been doing alot of AB testing. And it seems WebGLRenderer is also having issues so may not be WebGPU related although the cpu usage for WebGPU is higher. 

WebGPURenderer needs a colorSpace on the video texture or there is an error. If I remove the colorSpace for WebGLRenderer I get that colour issue again but it produces less cpu. So WebGLRenderer also needs SRGBColorSpace set on the texture.

I have an 153dev build that doesn't have any cpu or colour issues or requirement for SRGBColorSpace but I'm not sure which date that build was for. I picked one from the dev branch and has the colour issue also. 

https://danrossi.github.io/three-webgpu-renderer/tests/webgl_video_panorama_equirectangular.html

The tests aren't updated yet but if this colorspace is removed the cpu is lower. 

https://github.com/danrossi/three-webgpu-renderer/blob/master/tests/webgl_video_panorama_equirectangular_nosrgb.html#L123

**Comment by sunag** [comment]
> One thing I thought might be doing it comparing with the vanilla code. This map key is never get/set. So on each frame the shader is being setup again. The vanilla code does it external of the loop. Its because the shader string is being used as the map key. It should be a unique identifier of the shader string perhaps.

@danrossi Fixed this point.

**Comment by danrossi** [comment]
I don't want to spam the thread thanks. I did my own webgpu setup debug to compare with vanilla but can't find the issue, it's very significant cpu usage with video. The setups look exactly the same apart from buffers setup. If its a chrome problem I will report it. I think there is already a bug report for srgb when using SRGB8_ALPHA8 with webgl and was told it was unrelated but its clearly a problem with video textures.

**Comment by danrossi** [comment]
I've discovered with the WebGL performance issue throttling to 30fps with video callback alleviates the issue. But the cpu usage is still 13% with webgpu. Not sure what to report to Chrome yet. 

https://danrossi.github.io/three-webgpu-renderer/tests/webgpu_video_panorama_equirectangular_30fps.html

**Comment by danrossi** [comment]
I found the CPU usage issue it's another map key problem. The pipeline is being cleared and created on every loop.

**Comment by danrossi** [comment]
The latest commits lower the cpu, but now the gpu usage is 20%. WebGL usage is less. Is that intentional ? 

at 60fps its 20-40% gpu for one video texture

https://danrossi.github.io/three-webgpu-renderer/tests/webgpu_video_panorama_equirectangular.html

Webgl with chrome srgb fix is significantly lower in cpu still also. Webgpu slightly higher

https://danrossi.github.io/three-webgpu-renderer/tests/webgl_video_panorama_equirectangular_nosrgb.html

WebGPU at 30fps video callback is very low cpu and 2% gpu usage

https://danrossi.github.io/three-webgpu-renderer/tests/webgpu_video_panorama_equirectangular_30fps.html

**Comment by danrossi** [comment]
I've got an external project using the renderer changes with video texture showing the gpu usage https://danrossi.github.io/three-troika/test/webgpu.html

**Comment by danrossi** [comment]
Just an update of this. If I comment this, the gpu drops. The only difference to the vanilla gpu is two buffers added to the bind group. 

https://github.com/mrdoob/three.js/blob/dev/examples/jsm/renderers/webgpu/WebGPUBackend.js#L425

And this is reset on each frame so is setup again, so that code check is always empty.

https://github.com/mrdoob/three.js/blob/dev/examples/jsm/renderers/webgpu/WebGPUBackend.js#L437

**Comment by sunag** [comment]
🤔 Hmm... in the tests, the `attributeset` should be renewed with each render pass, and `bindingsData` should be unique per object. Maybe a PR can clarify better this.

**Comment by danrossi** [comment]
I've tried my best to identify why the gpu increases while the cpu is now low. It's when binding the group, after binding the group and before draw it's 30%, before its 2%. 

I tried to pull apart everything into a vanilla gpu render to see if there is something to report to Chrome but all the buffer setups are difficult to port over and bind group doesn't work. 

https://github.com/mrdoob/three.js/blob/6774f1e4aa52d10e0e14b017427bcf7e558b8d81/examples/jsm/renderers/webgpu/WebGPUBackend.js#L438

https://danrossi.github.io/three-webgpu-renderer/tests/webgpu_vanilla.html
https://github.com/danrossi/three-webgpu-renderer/blob/master/tests/webgpu_vanilla.html

The vanilla example has low resources but doesn't mimic everything three is doing in the bind group. 

https://danrossi.github.io/three-webgpu-renderer/tests/webgpu_basic.html

**Comment by danrossi** [comment]
I believe #26673 fixes video texture performance issues. Very low resources now. Closing thanks.

https://danrossi.github.io/three-webgpu-renderer/tests/webgpu_video_panorama_equirectangular.html

and using video frame callback

https://danrossi.github.io/three-webgpu-renderer/tests/webgpu_video_panorama_equirectangular_30fps.html

**Comment by mrdoob** [comment]
I think you misunderstood https://github.com/mrdoob/three.js/issues/26673?

**Comment by danrossi** [comment]
I didn't the cache fix in the past already corrected the cpu issue preventing setting things up per frame. But then it caused a high gpu issue. I've rebuilt and have low cpu and gpu resources now with video textures.

**Comment by mrdoob** [comment]
Hmmm... https://github.com/mrdoob/three.js/issues/26673 is a PR so it can't have fixed anything.
Maybe you mean https://github.com/mrdoob/three.js/pull/26556?

**Comment by danrossi** [comment]
Yeah I figured the cache things is still being worked on. Something wierd happened, maybe a Chrome update. I was not able to track down the last issue to confirm if it was a Chrome issue. But it was happening in the bind group. If I see anymore problems I'll update the other ticket and keep this closed. I have to see if I can properly integrate both renderers within an iife bundle now I figured how to do that.

**Comment by mch-outsight** [comment]
> Yeah I figured the cache things is still being worked on. Something wierd happened, maybe a Chrome update. I was not able to track down the last issue to confirm if it was a Chrome issue. But it was happening in the bind group. If I see anymore problems I'll update the other ticket and keep this closed. I have to see if I can properly integrate both renderers within an iife bundle now I figured how to do that.

Hi! We've observed performance issues using three.js on Chrome between Google Chrome v115 and v116, do you think it may be linked? We've submitted an issue [here](https://bugs.chromium.org/p/chromium/issues/detail?id=1475906&q=1475906&can=2).

**Comment by Mugen87** [comment]
No, this issue is about WebGPU not WebGL.

**Comment by danrossi** [comment]
I'm on 1.57.57 . Something randomly and dramatically changed the past month in chrome and I cannot replicate any WebGPU video texture issue now since three update. There was gpu usage of between 30-40%, now it's 3%. So will keep checking up on the PR issue for changes. With WebGL there is the unresolved Chrome cpu performance issue using srgb format with video textures, but they closed their ticket and considered it done. ? I guess I can use my vanilla WebGL code tests to reopen a ticket with them once I check it is still a problem.
